zerver/lib/markdown/__init__.py

```diff
@@ -574,14 +574,17 @@ def get_tweet_id(url: str) -> Optional[str]:
     return tweet_id_match.group("tweetid")
 
 
-class InlineHttpsProcessor(markdown.treeprocessors.Treeprocessor):
+class InlineImageProcessor(markdown.treeprocessors.Treeprocessor):
     def run(self, root: Element) -> None:
         # Get all URLs from the blob
         found_imgs = walk_tree(root, lambda e: e if e.tag == "img" else None)
         for img in found_imgs:
             url = img.get("src")
             assert url is not None
-            if urllib.parse.urlsplit(url).scheme != "http":
+            split_url = urllib.parse.urlsplit(url)
+            if (split_url.scheme == "https" and split_url.netloc in settings.ALLOWED_HOSTS) or (
+                split_url.scheme == "" and split_url.netloc == ""
+            ):
```

Comment:
This won’t work when `EXTERNAL_URI_SCHEME` is `"http://"`, `ALLOWED_HOSTS` has wildcards, or `url` has a port number. Maybe

```suggestion
            if (
                f"{split_url.scheme}://" == settings.EXTERNAL_URI_SCHEME
                and (
                    f".{split_url.netloc}".endswith(f".{settings.EXTERNAL_HOST}")
                    or split_url.netloc in settings.REALM_HOSTS.values()
                )
            ) or (split_url.scheme == "" and split_url.netloc == ""):
```

Reply:
Do we support running Zulip in a non-https environment? (not that we should break if we don't have to, I just recall hearing somewhere that we don't support that. But I could be misremembering)

Reply:
We do in development.  In production, we basically don't -- but we do support Zulip running behind a proxy where the protocol between the proxy and Zulip is HTTP (A lot of businesses seem to like this model, and we haven't decided to make it impossible to use that sort of security model that depends on the network between the two servers being secure).

Reply:
That makes sense. In cases where there's a reverse proxy, does `EXTERNAL_URI_SCHEME` refer to the scheme that zulip is speaking to the proxy, or the scheme that the proxy is speaking to the user?

If it's the former, I think that could mean that this logic is incorrect — we could have a server speaking HTTP to a proxy that speaks HTTPS to the user, and a link to `http://zulip.example.com/whatever.png` would be left as HTTP, even though we should be upgrading it to HTTPS.

Reply:
`EXTERNAL_HOST` and friends are used to generate links back to Zulip (E.g. in emails), and thus should always be the user-facing hostname (i.e. on the proxy), not the actual hostname for the Zulip server.